physical/postgresql/postgresql.go

```diff
@@ -96,7 +96,7 @@ func NewPostgreSQLBackend(conf map[string]string, logger log.Logger) (physical.B
 		get_query:    "SELECT value FROM " + quoted_table + " WHERE path = $1 AND key = $2",
 		delete_query: "DELETE FROM " + quoted_table + " WHERE path = $1 AND key = $2",
 		list_query: "SELECT key FROM " + quoted_table + " WHERE path = $1" +
-			"UNION SELECT DISTINCT substring(substr(path, length($1)+1) from '^.*?/') FROM " +
+			"UNION ALL SELECT DISTINCT substring(substr(path, length($1)+1) from '^.*?/') FROM " +
```

Comment:
Would there be any adverse impact here because of how `UNION ALL` doesn't de-duplicate results?

Reply:
Each branch of the union is a disjoint set.  The first set contains only the leaf values from the LIST target.  The second set contains all the intermediate paths lying below the LIST target.

We know they don't overlap because the second set's elements must end in "/" (as per the substring regexp), whereas no key coming from the first set may end in "/" (only provable by looking at the insert code, but it's true.)

The first set contains no dups due to the primary key.  The second set contains no dups because it uses DISTINCT.

Thus the `UNION` de-duping is unnecessary waste here.

Reply:
:+1: